crates/spk-storage/src/walker.rs

```diff
@@ -0,0 +1,1155 @@
+                                }
+                            }
+
+                            if self.emit_end_of_markers && !components.is_empty() {
+                                // Report the end of this build (after all the components in it)
+                                yield RepoWalkerItem::EndOfBuild(build.clone())
+                            }
+                        }
+
+                        if self.emit_end_of_markers && !builds.is_empty() {
+                            // Report the end of this version (after all its builds)
+                            yield RepoWalkerItem::EndOfVersion(package_version.clone())
+                        }
+                    }
+
+                    if self.emit_end_of_markers && !versions.is_empty() {
+                        // Report the end of this package (after all its versions)
+                        yield RepoWalkerItem::EndOfPackage(package.clone())
+                    }
+                }
+
+                if self.emit_end_of_markers && !packages.is_empty() {
+                    // Report the end of this repo (after all the packages in it)
+                    yield RepoWalkerItem::EndOfRepo(WalkedRepo { name: repo_name })
+                }
+            }
+        })
+    }
+}
+
+/// A builder for constructing a RepoWalker from various settings.
+///
+/// A default RepoWalker can made with:
+///
+///   let repo_walker_builder = RepoWalkerBuilder::new(repos);
+///   let repo_walker = builder.build();
+///
+/// That makes a RepoWalker that will: walk the given list of
+/// repositories to report on all packages, all versions, all builds
+/// that aren't /src or /deprecated builds, and will emit the package,
+/// version, and builds objects in sorted order (packages by name,
+/// versions by highest first, builds by digest). It will stop at
+/// builds, and will not walk components or files, and it won't emit
+/// "EndOf..." object markers. It will stop if it hits an error.
+///
+/// If the caller doesn't want a repository walked, they should not
+/// include it in the list given to the RepoWalkerBuilder's
+/// constructor.
+///
+/// Other walkers can be made by using the with_* methods on the
+/// RepoWalkerBuilder to configure it before calling build() and
+/// making the RepoWalker, e.g.
+///
+///   let repo_walker_builder = RepoWalkerBuilder::new(repos);
+///   let repo_walker = builder
+///         .try_with_package_equals(some_package)?
+///         .with_report_on_versions(true)
+///         .with_report_on_builds(true)
+///         .with_report_src_builds(true)
+///         .with_report_deprecated_builds(true)
+///         .with_report_embedded_builds(false)
+///         .with_report_on_components(true)
+///         .with_build_options_matching(host_options)
+///         .with_report_on_files(true)
+///         .with_file_path(some_file_path)
+///         .with_continue_on_error(true)
+///         .build();
```

Comment:
why not make these compiled examples?

this will also make sure that the markdown docs get proper syntax highlighting

```suggestion
///  ```no_run
///  let repo_walker_builder = RepoWalkerBuilder::new(repos);
///   let repo_walker = builder
///         .try_with_package_equals(some_package)?
///         .with_report_on_versions(true)
///         .with_report_on_builds(true)
///         .with_report_src_builds(true)
///         .with_report_deprecated_builds(true)
///         .with_report_embedded_builds(false)
///         .with_report_on_components(true)
///         .with_build_options_matching(host_options)
///         .with_report_on_files(true)
///         .with_file_path(some_file_path)
///         .with_continue_on_error(true)
///         .build();
/// ```
```

Reply:
This needed a rebase to pick up the changes in main that it had to turn these into doctests.